posts/2023-06-07-kotlin-ktor-keycloak.adoc

```diff
@@ -0,0 +1,322 @@
+            ) {
+                withAudience("account")
+                withClaim("azp", "luna-vacation-bot")
+            }
+            challenge { _, _ ->
+                call.respond(HttpStatusCode.Unauthorized, "Token is not valid or has expired")
+            }
+            validate { credential ->
+                validateCredential(credential, issuer)
+            }
+        }
+    }
+----
+If the authentication fails it returns the `challenge`, in this case a 401 Unauthorized with an error message.
+
+The `validateCredentials` function is simply validating that the token has not expired, after all other validations have passed.
+
+[source,kotlin]
+----
+fun validateCredential(credential: JWTCredential, issuer: String): JWTPrincipal? {
+    if (credential.expiresAt?.after(Date()) == true
+    ) {
+        return JWTPrincipal(credential.payload)
+    }
+
+    return null
+}
+----
+
+==== Test the API
+We can test the api using curl.
+If we secured the endpoint properly calling the `imionvacation` should return an error.
+
+[source,commandline]
+----
+$ curl localhost:4040/imionvacation
+Token is not valid or has expired
+----
+
+As the diagram at the beginning of this post explained we need to request a token to Keycloak that we send together with the request, and for that we impersonate the `luna-vacation-bot`:
+
+[source,commandline]
+----
+curl -X post 'http://localhost:8080/realms/lunarealm/protocol/openid-connect/token' \
+--header 'Content-Type: application/x-www-form-urlencoded' \
+--data-urlencode 'grant_type=client_credentials' \
+--data-urlencode 'client_id=luna-vacation-bot' \
+--data-urlencode 'client_secret=<replace-by-secret>'
+----
+
+Grab the `access_token` part of the response and add it to the request header as a Bearer token:
+
+[source,commandline]
+----
+$ curl 'localhost:4040/imionvacation' \
+--header 'Authorization: Bearer eyJhbGciOiJ...QBhNiX6w'
+----
+
+And now we get a result:
+
+[source,commandline]
+----
+Today you have to work
+----
+
+
+=== Create luna-vacation-bot
+
+For the bot app we are going to use Intellij (or your favorite IDE) and create a new Kotlin application, with gradle and the latest JVM you have available (I'm using JDK 17). After the project is created you can remove the test folder and the test dependencies.
+
+==== Integrate with luna-vacation-api
+
+For the http request let's use `ktor-client` and add the necessary dependencies to `build.gradle.kts`:
+
+[source,kotlin]
+----
+implementation("io.ktor:ktor-client-core:2.3.1")
+implementation("io.ktor:ktor-client-cio:2.3.1")
+implementation("io.ktor:ktor-client-serialization:2.3.1")
+implementation("io.ktor:ktor-client-content-negotiation:2.3.1")
+implementation("io.ktor:ktor-serialization-kotlinx-json:2.3.1")
+----
+
+To create an HttpClient we need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization,
+where we will ignore the json fields that we are not interested in.
```

Comment:
This sentence is very long... Would this be a correct modification?
```
To create an HttpClient we need to add the json `ContentNegotiation` and the flag `ignoreUnknownKeys` for the response deserialization. This will ignore the json fields that we are not interested in.
```